### PR TITLE
Control first downsample stride in ResNet

### PR DESCRIPTION
**Pull request by jiqing-feng**
Hi @ArthurZucker 

Relate to [25856](https://github.com/huggingface/transformers/pull/25856). I added a new parameter in config to control the stride for the first bottleneck layer in stages. Would you please help me review it? Thx!

### CODE REVIEW

**Review by rafaelpadilla — CHANGES_REQUESTED** [review]
A new configuration parameter is being suggested in the ResNet config to attend the needs of a new model. It was done to preserve backward compatibility by default. :) 

However, the name of the suggested config parameter should be more intuitive to leverage other models. Could you please, make this change?

**Review by ArthurZucker — COMMENTED** [review]
Thanks for opening the follow up PR 🤗

**Comment by jiqing-feng** [comment]
Hi @ArthurZucker @rafaelpadilla .

Thanks for your advice. The resnet backbone which downsamples in bottleneck layer can be found in [resnet](https://huggingface.co/Jiqing/resnet-backbone-downsample_in_bottleneck).

File: src/transformers/models/resnet/modeling_resnet.py [hunk 1]
```diff
@@ -123,7 +123,14 @@ class ResNetBasicLayer(nn.Module):
...
     """
 
-    def __init__(self, in_channels: int, out_channels: int, stride: int = 1, activation: str = "relu"):
+    def __init__(
+        self,
+        in_channels: int,
+        out_channels: int,
+        stride: int = 1,
+        activation: str = "relu",
+        downsample_in_bottleneck: bool = False,
```
Comment (rafaelpadilla) [inline src/transformers/models/resnet/modeling_resnet.py:132]: I see the purpose of `downsample_in_bottleneck` here: it prevents one from distinguishing between ResNetConvLayer and ResNetBasicLayer.
So, I think it is valuable to include a docstring here, and clarify why `downsample_in_bottleneck` isn't used.

File: src/transformers/models/resnet/modeling_resnet.py [hunk 2]
```diff
@@ -153,7 +160,13 @@ class ResNetBottleNeckLayer(nn.Module):
...
 
     def __init__(
-        self, in_channels: int, out_channels: int, stride: int = 1, activation: str = "relu", reduction: int = 4
+        self,
+        in_channels: int,
+        out_channels: int,
+        stride: int = 1,
+        activation: str = "relu",
+        reduction: int = 4,
+        downsample_in_bottleneck: bool = False,
```
Comment (rafaelpadilla) [inline src/transformers/models/resnet/modeling_resnet.py:169]: Including a docstring here would be very valuable.

File: src/transformers/models/resnet/modeling_resnet.py (hunk 1 above)
Comment (rafaelpadilla) [inline src/transformers/models/resnet/modeling_resnet.py:132]: I see the purpose of `downsample_in_bottleneck` here: it prevents one from distinguishing between `ResNetBottleNeckLayer` and `ResNetBasicLayer`. But it is not being use in this class.
So, I think it is valuable to include a docstring here, and clarify why `downsample_in_bottleneck` isn't used.

**Review by rafaelpadilla — APPROVED** [review]
Looks good to me. 
But including docstring in the constructors of `ResNetBottleNeckLayer` and `ResNetBasicLayer` seems important to me, as the new parameter `downsample_in_bottleneck` is introduced.

**Review by ArthurZucker — COMMENTED** [review]
LGTM appart for the arg that is not used. Let's have explicit and useful args, doing a if else is okay to chose the layer and more understandable

**Comment by jiqing-feng** [comment]
> LGTM appart for the arg that is not used. Let's have explicit and useful args, doing a if else is okay to chose the layer and more understandable

Sure!

File: src/transformers/models/resnet/modeling_resnet.py (hunk 1 above)
Comment (rafaelpadilla) [reply src/transformers/models/resnet/modeling_resnet.py]: Agree! It's cleaner.

File: src/transformers/models/resnet/modeling_resnet.py [hunk 3]
```diff
@@ -194,11 +203,24 @@ def __init__(
...
+                    downsample_in_bottleneck=config.downsample_in_bottleneck,
+                ),
+                *[layer(out_channels, out_channels, activation=config.hidden_act) for _ in range(depth - 1)],
+            )
+        else:
+            self.layers = nn.Sequential(
+                # downsampling is done in the first layer with stride of 2
+                layer(in_channels, out_channels, stride=stride, activation=config.hidden_act),
+                *[layer(out_channels, out_channels, activation=config.hidden_act) for _ in range(depth - 1)],
+            )
```
Comment (ArthurZucker) [inline src/transformers/models/resnet/modeling_resnet.py:223]: No as I said before the only layer that is different here is the call. Something like 
```suggestion
        if config.layer_type == "bottleneck":
                first_layer = layer(in_channels,out_channels,stride=stride,activation=config.hidden_act,downsample_in_bottleneck=config.downsample_in_bottleneck)
        else:
                first_layer =  layer(in_channels, out_channels, stride=stride, activation=config.hidden_act)
        self.layers =  nn.Sequential(first_layer,*[layer(out_channels, out_channels, activation=config.hidden_act) for _ in range(depth - 1)])
```

Comment (ArthurZucker) [reply src/transformers/models/resnet/modeling_resnet.py:223]: something like this should be enough.

**Review by ArthurZucker — COMMENTED** [review]
Sorry, this should be the last iteration!

File: src/transformers/models/resnet/modeling_resnet.py (hunk 3 above)
Comment (jiqing-feng) [reply src/transformers/models/resnet/modeling_resnet.py]: Sure!

**Review by ArthurZucker — APPROVED** [review]
thanks for this round! merging 😉

**Pull request merged**